src/gasman.h

```diff
@@ -293,18 +298,71 @@ EXPORT_INLINE UInt SIZE_BAG_CONTENTS(const void *ptr) {
 **  the application  must inform {\Gasman}  that it  has changed  the bag, by
 **  calling 'CHANGED_BAG(old)' in the above example (see "CHANGED_BAG").
 */
+#ifdef HPCGAP
+EXPORT_INLINE int ReadCheck(Bag bag)
```

Comment:
I am also generally wondering what the performance impact of this PR is... Like, with something trivial, as running testinstall with and without guards, how does this change the timings (that's a very crude and bad benchmark, of course, but better than nothing)

Reply:
I removed the `PURE_FUNC` annotations, because it made no noticeable difference on performance.

Reply:
Oh, and when running `tst/testinstall.g` without having a full GC after each test (to better simulate normal behavior), guard overhead is currently down to a bit less than 10%. There's too much noise to pin the overhead down better, but it's actually an improvement over the original Ward version (where it was around 15%). I suspect that this is largely because modern C compilers do a much better job at optimizing away unnecessary checks than the fairly basic dataflow analysis of Ward did (which, for starters, didn't allow for interprocedural analysis).

Reply:
So did you have a look at the disassembly to see whether the compiled hoisted guard checks out of loops at all?

Reply:
So did you have a look at the disassembly to see whether the compiled hoisted guard checks out of loops at all?

Reply:
I just checked myself, by looking at `EqRange`, which is rather simple and ideally only would need to `ReadGuard` calls, but naively requires six. And sadly, what I am seeing are the full six calls. Here is disassembly in plain GAP with your PR:
```asm
$ otool -p _EqRange -x -V ./gap
./gap:
(__TEXT,__text) section
_EqRange:
0000000100107940	pushq	%rbp
0000000100107941	movq	%rsp, %rbp
0000000100107944	movq	(%rdi), %rax
0000000100107947	movq	(%rsi), %rcx
000000010010794a	movq	(%rcx), %rdx
000000010010794d	xorq	(%rax), %rdx
0000000100107950	cmpq	$0x3, %rdx
0000000100107954	ja	0x100107977
0000000100107956	movq	0x8(%rcx), %rdx
000000010010795a	xorq	0x8(%rax), %rdx
000000010010795e	cmpq	$0x3, %rdx
0000000100107962	ja	0x10010797b
0000000100107964	movq	0x10(%rcx), %rcx
0000000100107968	xorq	0x10(%rax), %rcx
000000010010796c	xorl	%eax, %eax
000000010010796e	cmpq	$0x4, %rcx
0000000100107972	setb	%al
0000000100107975	popq	%rbp
0000000100107976	retq
0000000100107977	xorl	%eax, %eax
0000000100107979	popq	%rbp
000000010010797a	retq
000000010010797b	xorl	%eax, %eax
000000010010797d	popq	%rbp
000000010010797e	retq
000000010010797f	nop
```

And here is what I get in HPC-GAP with your PR:
```asm
$ otool -p _EqRange -x -V ./gap
./gap:
(__TEXT,__text) section
_EqRange:
0000000100156730	pushq	%rbp
0000000100156731	movq	%rsp, %rbp
0000000100156734	pushq	%r15
0000000100156736	pushq	%r14
0000000100156738	pushq	%rbx
0000000100156739	pushq	%rax
000000010015673a	movq	%rsi, %r14
000000010015673d	movq	%rdi, %r15
0000000100156740	movq	0x8(%rdi), %rax
0000000100156744	testq	%rax, %rax
0000000100156747	je	0x10015676b
0000000100156749	movq	%rbp, %rcx
000000010015674c	andq	$-0x100000, %rcx
0000000100156753	cmpq	%rcx, 0x28(%rax)
0000000100156757	je	0x10015676b
0000000100156759	movslq	(%rcx), %rcx
000000010015675c	cmpb	$0x0, 0x50(%rax,%rcx)
0000000100156761	jne	0x10015676b
0000000100156763	movq	%r15, %rdi
0000000100156766	callq	_HandleReadGuardError
000000010015676b	movq	(%r15), %rax
000000010015676e	movq	(%rax), %rbx
0000000100156771	movq	0x8(%r14), %rax
0000000100156775	testq	%rax, %rax
0000000100156778	je	0x10015679c
000000010015677a	movq	%rbp, %rcx
000000010015677d	andq	$-0x100000, %rcx
0000000100156784	cmpq	%rcx, 0x28(%rax)
0000000100156788	je	0x10015679c
000000010015678a	movslq	(%rcx), %rcx
000000010015678d	cmpb	$0x0, 0x50(%rax,%rcx)
0000000100156792	jne	0x10015679c
0000000100156794	movq	%r14, %rdi
0000000100156797	callq	_HandleReadGuardError
000000010015679c	movq	(%r14), %rax
000000010015679f	xorq	(%rax), %rbx
00000001001567a2	cmpq	$0x3, %rbx
00000001001567a6	ja	0x100156885
00000001001567ac	movq	0x8(%r15), %rax
00000001001567b0	testq	%rax, %rax
00000001001567b3	je	0x1001567d7
00000001001567b5	movq	%rbp, %rcx
00000001001567b8	andq	$-0x100000, %rcx
00000001001567bf	cmpq	%rcx, 0x28(%rax)
00000001001567c3	je	0x1001567d7
00000001001567c5	movslq	(%rcx), %rcx
00000001001567c8	cmpb	$0x0, 0x50(%rax,%rcx)
00000001001567cd	jne	0x1001567d7
00000001001567cf	movq	%r15, %rdi
00000001001567d2	callq	_HandleReadGuardError
00000001001567d7	movq	(%r15), %rax
00000001001567da	movq	0x8(%rax), %rbx
00000001001567de	movq	0x8(%r14), %rax
00000001001567e2	testq	%rax, %rax
00000001001567e5	je	0x100156809
00000001001567e7	movq	%rbp, %rcx
00000001001567ea	andq	$-0x100000, %rcx
00000001001567f1	cmpq	%rcx, 0x28(%rax)
00000001001567f5	je	0x100156809
00000001001567f7	movslq	(%rcx), %rcx
00000001001567fa	cmpb	$0x0, 0x50(%rax,%rcx)
00000001001567ff	jne	0x100156809
0000000100156801	movq	%r14, %rdi
0000000100156804	callq	_HandleReadGuardError
0000000100156809	movq	(%r14), %rax
000000010015680c	xorq	0x8(%rax), %rbx
0000000100156810	cmpq	$0x3, %rbx
0000000100156814	ja	0x100156885
0000000100156816	movq	0x8(%r15), %rax
000000010015681a	testq	%rax, %rax
000000010015681d	je	0x100156841
000000010015681f	movq	%rbp, %rcx
0000000100156822	andq	$-0x100000, %rcx
0000000100156829	cmpq	%rcx, 0x28(%rax)
000000010015682d	je	0x100156841
000000010015682f	movslq	(%rcx), %rcx
0000000100156832	cmpb	$0x0, 0x50(%rax,%rcx)
0000000100156837	jne	0x100156841
0000000100156839	movq	%r15, %rdi
000000010015683c	callq	_HandleReadGuardError
0000000100156841	movq	(%r15), %rax
0000000100156844	movq	0x10(%rax), %rbx
0000000100156848	movq	0x8(%r14), %rax
000000010015684c	testq	%rax, %rax
000000010015684f	je	0x100156873
0000000100156851	movq	%rbp, %rcx
0000000100156854	andq	$-0x100000, %rcx
000000010015685b	cmpq	%rcx, 0x28(%rax)
000000010015685f	je	0x100156873
0000000100156861	movslq	(%rcx), %rcx
0000000100156864	cmpb	$0x0, 0x50(%rax,%rcx)
0000000100156869	jne	0x100156873
000000010015686b	movq	%r14, %rdi
000000010015686e	callq	_HandleReadGuardError
0000000100156873	movq	(%r14), %rax
0000000100156876	xorq	0x10(%rax), %rbx
000000010015687a	xorl	%eax, %eax
000000010015687c	cmpq	$0x4, %rbx
0000000100156880	setb	%al
0000000100156883	jmp	0x100156887
0000000100156885	xorl	%eax, %eax
0000000100156887	addq	$0x8, %rsp
000000010015688b	popq	%rbx
000000010015688c	popq	%r14
000000010015688e	popq	%r15
0000000100156890	popq	%rbp
0000000100156891	retq
0000000100156892	nopw	%cs:(%rax,%rax)
000000010015689c	nopl	(%rax)
```


Of course that particular function could (should?) easily be optimized, in fact it could simply call `memcmp`. I only chose it because it's easy to see with a glance what is going on, even if one is not familiar with assembly.

Reply:
Just to clarify: the above code is from Apple clang version 11.0.0 (clang-1100.0.33.17) compiled with our default -O2 optimization, and even the plain GAP version obviously is not optimal. @rbehrends pointed out to me that GCC does a better job. Indeed, here is what I got with GCC 9.2.0.

Plain GAP:
```asm
$ otool -p _EqRange -t -V ./gap
./gap:
(__TEXT,__text) section
_EqRange:
00000001000e5040	movq	(%rdi), %rcx
00000001000e5043	xorl	%r8d, %r8d
00000001000e5046	movq	(%rsi), %rsi
00000001000e5049	movq	(%rcx), %rax
00000001000e504c	movq	(%rsi), %rdx
00000001000e504f	sarq	$0x2, %rax
00000001000e5053	sarq	$0x2, %rdx
00000001000e5057	cmpq	%rax, %rdx
00000001000e505a	je	0x1000e5060
00000001000e505c	movq	%r8, %rax
00000001000e505f	retq
00000001000e5060	movq	0x8(%rsi), %rdx
00000001000e5064	movq	0x8(%rcx), %rax
00000001000e5068	sarq	$0x2, %rdx
00000001000e506c	sarq	$0x2, %rax
00000001000e5070	cmpq	%rax, %rdx
00000001000e5073	jne	0x1000e505c
00000001000e5075	movq	0x10(%rsi), %rdx
00000001000e5079	xorl	%r8d, %r8d
00000001000e507c	movq	0x10(%rcx), %rax
00000001000e5080	sarq	$0x2, %rdx
00000001000e5084	sarq	$0x2, %rax
00000001000e5088	cmpq	%rax, %rdx
00000001000e508b	sete	%r8b
00000001000e508f	movq	%r8, %rax
00000001000e5092	retq
00000001000e5093	nopw	%cs:(%rax,%rax)
00000001000e509d	nopl	(%rax)
```

And with HPC-GAP, this (clearly much more complicated, but with only two guard `HandleReadGuardError` calls):
```asm
$ otool -p _EqRange -t -V ./gap
./gap:
(__TEXT,__text) section
_EqRange:
0000000100129d30	pushq	%rbp
0000000100129d31	movq	%rsp, %rbp
0000000100129d34	pushq	%r13
0000000100129d36	movq	%rsi, %r13
0000000100129d39	pushq	%r12
0000000100129d3b	movq	%rdi, %r12
0000000100129d3e	pushq	%rbx
0000000100129d3f	subq	$0x8, %rsp
0000000100129d43	movq	0x8(%rdi), %rax
0000000100129d47	testq	%rax, %rax
0000000100129d4a	je	0x100129d73
0000000100129d4c	movq	%rbp, %rdx
0000000100129d4f	andq	$-0x100000, %rdx
0000000100129d56	cmpq	%rdx, 0x28(%rax)
0000000100129d5a	je	0x100129d73
0000000100129d5c	movslq	(%rdx), %rdx
0000000100129d5f	cmpb	$0x0, 0x50(%rax,%rdx)
0000000100129d64	jne	0x100129d73
0000000100129d66	callq	_ExtendedGuardCheck
0000000100129d6b	testl	%eax, %eax
0000000100129d6d	je	0x100129e15
0000000100129d73	movq	(%r12), %rbx
0000000100129d77	movq	0x8(%r13), %rax
0000000100129d7b	movq	(%rbx), %r12
0000000100129d7e	sarq	$0x2, %r12
0000000100129d82	testq	%rax, %rax
0000000100129d85	je	0x100129dad
0000000100129d87	movq	%rbp, %rdx
0000000100129d8a	andq	$-0x100000, %rdx
0000000100129d91	cmpq	%rdx, 0x28(%rax)
0000000100129d95	je	0x100129dad
0000000100129d97	movslq	(%rdx), %rdx
0000000100129d9a	cmpb	$0x0, 0x50(%rax,%rdx)
0000000100129d9f	jne	0x100129dad
0000000100129da1	movq	%r13, %rdi
0000000100129da4	callq	_ExtendedGuardCheck
0000000100129da9	testl	%eax, %eax
0000000100129dab	je	0x100129e0d
0000000100129dad	movq	(%r13), %rdx
0000000100129db1	xorl	%r8d, %r8d
0000000100129db4	movq	(%rdx), %rax
0000000100129db7	sarq	$0x2, %rax
0000000100129dbb	cmpq	%r12, %rax
0000000100129dbe	je	0x100129dd0
0000000100129dc0	addq	$0x8, %rsp
0000000100129dc4	movq	%r8, %rax
0000000100129dc7	popq	%rbx
0000000100129dc8	popq	%r12
0000000100129dca	popq	%r13
0000000100129dcc	popq	%rbp
0000000100129dcd	retq
0000000100129dce	nop
0000000100129dd0	movq	0x8(%rdx), %rcx
0000000100129dd4	movq	0x8(%rbx), %rax
0000000100129dd8	sarq	$0x2, %rcx
0000000100129ddc	sarq	$0x2, %rax
0000000100129de0	cmpq	%rax, %rcx
0000000100129de3	jne	0x100129dc0
0000000100129de5	movq	0x10(%rbx), %rax
0000000100129de9	xorl	%r8d, %r8d
0000000100129dec	movq	0x10(%rdx), %rdx
0000000100129df0	sarq	$0x2, %rax
0000000100129df4	sarq	$0x2, %rdx
0000000100129df8	cmpq	%rax, %rdx
0000000100129dfb	sete	%r8b
0000000100129dff	addq	$0x8, %rsp
0000000100129e03	popq	%rbx
0000000100129e04	movq	%r8, %rax
0000000100129e07	popq	%r12
0000000100129e09	popq	%r13
0000000100129e0b	popq	%rbp
0000000100129e0c	retq
0000000100129e0d	movq	%r13, %rdi
0000000100129e10	callq	_HandleReadGuardError
0000000100129e15	movq	%r12, %rdi
0000000100129e18	callq	_HandleReadGuardError
0000000100129e1d	nopl	(%rax)
```